gpflow/misc.py

```diff
@@ -251,5 +251,57 @@ def _get_tensor_safe(name, index, graph):
     except KeyError:
         return None
 
+
+def swap_final_dims(X: tf.Tensor):
+    """
+    equivalent to np.swapaxes(X, -2, -1)
+    """
+    K = tf.rank(X)
+    num_broadcasting_dims = tf.reduce_max([K - 2, 0])
+    perm = tf.concat([tf.range(num_broadcasting_dims), [K - 1, K - 2]], 0)
+    return tf.transpose(X, perm)
+
+
+def slice_final_dim(X: tf.Tensor, sl: slice, dim: int):
+    """
+    Takes the slice sl from the final dimension of X
+    If the slice is the whole of the final dimension (i.e. sl=slice(0, dim)) then there is nothing to do
+    NB tensorflow does not support slicing with step!=1, so raise error if this is attempted
+    """
+    if (sl.start is None) and (sl.step is None) and (sl.stop == dim):  # nothing to do
```

Comment:
Simply: `sl == slice(None, dim, None)`

Reply:
actually it was this line that was breaking some things. It seems there are situations where the slice is as big as input_dim, but it still needs to do something. Removing so now more tests pass, though it seems we're doing a bit of unnecessary slicing here (and in the current master, too)